golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,510 @@
+        .bind(component.component_id)
+        .bind(component.name.clone())
+        .execute(&mut *transaction)
+        .await?;
+
+        sqlx::query(
+            r#"
+              INSERT INTO component_versions
+                (component_id, version, size, user_component, protected_component, protector_version, metadata)
+              VALUES
+                ($1, $2, $3, $4, $5, $6, $7)
+              ON CONFLICT (component_id, version) DO UPDATE
+              SET size = $3,
+                  user_component = $4,
+                  protected_component = $5,
+                  protector_version = $6,
+                  metadata = $7
+               "#,
+        )
+            .bind(component.component_id)
+            .bind(component.version)
+            .bind(component.size)
+            .bind(component.user_component.clone())
+            .bind(component.protected_component.clone())
+            .bind(component.protector_version)
+            .bind(component.metadata.clone())
+            .execute(&mut *transaction)
+            .await?;
+
+        transaction.commit().await?;
+
+        Ok(())
+    }
+
+    async fn get(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Vec<ComponentRecord>, RepoError> {
+        sqlx::query_as::<_, ComponentRecord>(
+            "SELECT components.namespace, components.name, component_versions.* FROM components JOIN component_versions ON components.component_id = component_versions.component_id WHERE components.namespace = $1 AND component_versions.component_id = $2"
+        )
+            .bind(namespace)
+            .bind(component_id)
+            .fetch_all(self.db_pool.deref())
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn get_all(&self, namespace: &str) -> Result<Vec<ComponentRecord>, RepoError> {
+        sqlx::query_as::<_, ComponentRecord>("SELECT components.namespace, components.name, component_versions.* FROM components JOIN component_versions ON components.component_id = component_versions.component_id WHERE components.namespace = $1")
+            .bind(namespace)
+            .fetch_all(self.db_pool.deref())
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn get_latest_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Option<ComponentRecord>, RepoError> {
+        sqlx::query_as::<_, ComponentRecord>(
+            "SELECT components.namespace, components.name, component_versions.* FROM components JOIN component_versions ON components.component_id = component_versions.component_id WHERE components.namespace = $1 AND component_versions.component_id = $2  ORDER BY version DESC LIMIT 1",
+        ).bind(namespace)
+            .bind(component_id)
+            .fetch_optional(self.db_pool.deref())
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn get_by_version(
+        &self,
+        namespace: &str,
```

Comment:
namespace is never needed if you have `component_id` for query

Reply:
May be namespace should be a key as well even if technically component_id is unique.